File: projects/design-system/scripts-custom/validate-token-syntax.py
#!/usr/bin/env python3
"""
Quick Token Syntax Validation

This script performs fast validation of token reference syntax:
- Checks for plural forms in token references (colors → color)
- Validates inversity suffix requirements for color tokens
- Identifies common syntax issues after refactoring

USAGE:
    python3 scripts-custom/validate-token-syntax.py

Run from project root for quick validation after changes.
"""

import json
import os
import re
from collections import defaultdict
from pathlib import Path

def load_all_tokens():
    """Load all tokens from JSON files."""
    tokens = {}
    themes_dir = Path('src/lib/themes')
    
    for json_file in themes_dir.rglob('*.json'):
        if json_file.name.startswith('$') or '_ignore-in-ds' in str(json_file):
            continue
            
        try:
            with open(json_file, 'r', encoding='utf-8') as f:
                data = json.load(f)
                tokens[str(json_file)] = data
        except json.JSONDecodeError as e:
            print(f"JSON error in {json_file}: {e}")
        except Exception as e:
            print(f"Error loading {json_file}: {e}")
    
    return tokens

def extract_references(obj, path=""):
    """Extract all token references from a JSON object."""
    references = []
    
    if isinstance(obj, dict):
        for key, value in obj.items():
            current_path = f"{path}.{key}" if path else key
            if key == "$value" and isinstance(value, str) and value.startswith("{") and value.endswith("}"):
                # Extract the reference
                ref = value[1:-1]  # Remove { and }
                references.append((current_path, ref))
            elif isinstance(value, (dict, list)):
                references.extend(extract_references(value, current_path))
    elif isinstance(obj, list):
        for i, item in enumerate(obj):
            current_path = f"{path}[{i}]"
            references.extend(extract_references(item, current_path))
    
    return references

def check_references():
    """Check all token references for validity."""
    tokens = load_all_tokens()
    all_references = []
    
    # Extract all references
    for file_path, data in tokens.items():
        refs = extract_references(data)
        for path, ref in refs:
            all_references.append((file_path, path, ref))
    
    print(f"Found {len(all_references)} token references")
    
    # Check for common issues
    issues = []
    
    for file_path, path, ref in all_references:
        # Check for plural forms in references
        if re.search(r'\b(colors|components|semantics|primitives)\b', ref):
            issues.append(f"Plural reference in {file_path} at {path}: {ref}")
        
        # Check for missing inversity suffixes in color references
        if 'ob.s.color' in ref and not ref.endswith(('.inversity-normal', '.inversity-flipped', '.inversity-flipped-alpha')):
            # Check if it's a theme reference or needs inversity suffix
            if not any(x in ref for x in ['theme-configuration', 'static', 'brand']):
                issues.append(f"Missing inversity suffix in {file_path} at {path}: {ref}")
    
    if issues:
        print(f"\nFound {len(issues)} potential issues:")
        for issue in issues:
            print(f"  {issue}")
    else:
        print("\nNo major issues found!")
    
    return len(issues) == 0

if __name__ == "__main__":
    check_references()

File: projects/design-system/scripts-custom/validate-token-syntax.js
#!/usr/bin/env node
/**
 * Quick Token Syntax Validation
 * 
 * This script performs fast validation of token reference syntax:
 * - Checks for plural forms in token references (colors → color)
 * - Validates inversity suffix requirements for color tokens
 * - Identifies common syntax issues after refactoring
 * 
 * USAGE:
 *     node scripts-custom/validate-token-syntax.js
 * 
 * Run from project root for quick validation after changes.
 */

const fs = require('fs');
const path = require('path');
const { glob } = require('glob');

/**
 * Load all tokens from JSON files
 */
async function loadAllTokens() {
    const tokens = {};
    const themesDir = 'src/lib/themes';
    
    try {
        const jsonFiles = await glob('**/*.json', { cwd: themesDir });
        
        for (const jsonFile of jsonFiles) {
            const fullPath = path.join(themesDir, jsonFile);
            
            if (path.basename(jsonFile).startsWith('$') || jsonFile.includes('_ignore-in-ds')) {
                continue;
            }
            
            try {
                const content = fs.readFileSync(fullPath, 'utf-8');
                const data = JSON.parse(content);
                tokens[fullPath] = data;
            } catch (error) {
                if (error instanceof SyntaxError) {
                    console.log(`JSON error in ${jsonFile}: ${error.message}`);
                } else {
                    console.log(`Error loading ${jsonFile}: ${error.message}`);
                }
            }
        }
    } catch (error) {
        console.log(`Error scanning themes directory: ${error.message}`);
    }
    
    return tokens;
}

/**
 * Extract all token references from a JSON object
 */
function extractReferences(obj, currentPath = '') {
    const references = [];
    
    if (typeof obj === 'object' && obj !== null) {
        if (Array.isArray(obj)) {
            for (let i = 0; i < obj.length; i++) {
                const itemPath = `${currentPath}[${i}]`;
                references.push(...extractReferences(obj[i], itemPath));
            }
        } else {
            for (const [key, value] of Object.entries(obj)) {
                const keyPath = currentPath ? `${currentPath}.${key}` : key;
                
                if (key === '$value' && typeof value === 'string' && value.startsWith('{') && value.endsWith('}')) {
                    // Extract the reference
                    const ref = value.slice(1, -1); // Remove { and }
                    references.push([keyPath, ref]);
                } else if (typeof value === 'object' && value !== null) {
                    references.push(...extractReferences(value, keyPath));
                }
            }
        }
    }
    
    return references;
}

/**
 * Check all token references for validity
 */
async function checkReferences() {
    const tokens = await loadAllTokens();
    const allReferences = [];
    
    // Extract all references
    for (const [filePath, data] of Object.entries(tokens)) {
        const refs = extractReferences(data);
        for (const [tokenPath, ref] of refs) {
            allReferences.push([filePath, tokenPath, ref]);
        }
    }
    
    console.log(`Found ${allReferences.length} token references`);
    
    // Check for common issues
    const issues = [];
    
    for (const [filePath, tokenPath, ref] of allReferences) {
        // Check for plural forms in references
        if (/\b(colors|components|semantics|primitives)\b/.test(ref)) {
            issues.push(`Plural reference in ${filePath} at ${tokenPath}: ${ref}`);
        }
        
        // Check for missing inversity suffixes in color references
        if (ref.includes('ob.s.color') && 
            !ref.endsWith('.inversity-normal') && 
            !ref.endsWith('.inversity-flipped') && 
            !ref.endsWith('.inversity-flipped-alpha')) {
            
            // Check if it's a theme reference or needs inversity suffix
            const skipPatterns = ['theme-configuration', 'static', 'brand'];
            if (!skipPatterns.some(pattern => ref.includes(pattern))) {
                issues.push(`Missing inversity suffix in ${filePath} at ${tokenPath}: ${ref}`);
            }
        }
    }
    
    if (issues.length > 0) {
        console.log(`\nFound ${issues.length} potential issues:`);
        for (const issue of issues) {
            console.log(`  ${issue}`);
        }
    } else {
        console.log('\nNo major issues found!');
    }
    
    return issues.length === 0;
}

/**
 * Main function
 */
async function main() {
    try {
        const success = await checkReferences();
        if (!success) {
            process.exit(1);
        }
    } catch (error) {
        console.error('Error:', error);
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

module.exports = { loadAllTokens, extractReferences, checkReferences, main };
